pkg/packagekit/updates.jsx

```diff
@@ -350,6 +354,108 @@ const UpdatesList = ({ updates }) => {
     );
 };
 
+class RestartServices extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            dialogError: undefined,
+        };
+
+        this.dialogErrorSet = this.dialogErrorSet.bind(this);
+        this.dialogErrorDismiss = this.dialogErrorDismiss.bind(this);
+        this.restart = this.restart.bind(this);
+    }
+
+    dialogErrorSet(text, detail) {
+        this.setState({ dialogError: text, dialogErrorDetail: detail });
+    }
+
+    dialogErrorDismiss() {
+        this.setState({ dialogError: undefined });
+    }
+
+    restart() {
+        // make sure cockpit package is the last to restart
+        const daemons = this.props.daemons.sort((a, b) => {
+            if (a.includes("cockpit") && b.includes("cockpit"))
+                return 0;
+            if (a.includes("cockpit"))
+                return 1;
+            if (b.includes("cockpit"))
+                return -1;
+        });
+        Promise.all(daemons.map(service => cockpit.dbus("org.freedesktop.systemd1", { superuser: "try" }).call("/org/freedesktop/systemd1", "org.freedesktop.systemd1.Manager", "RestartUnit", [service + ".service", "fail"]))) // TODO be smarter than just concatenating ".service"
+                .then(() => {
+                    this.props.onValueChanged({ restartableServices: { reboot: [], daemons: [], manual: [] } });
+                    if (this.props.state === "updateSuccess")
+                        this.props.loadUpdates();
+                    this.props.close();
+                })
+                .catch(ex => this.dialogErrorSet("RestartUnit: failed to restart a unit", ex.message));
```

Comment:
state update? You mean call loadUpdates(). But if that is called, user would have no time to read the error message. Unless we do the error handling thru patternfly toast Alert notifications.
So generally, I think the whole system of error handling is not really good at updates page. It needs bigger change, like rewriting an error handling so all errors are showed in unified way. For example thru toast notifications like the one which just got implemented for machines page. That would fix the issue here that both an error message would be visible, and the state would update. I could also try to solve https://github.com/cockpit-project/cockpit/issues/11222 .
WDYT?

Reply:
I didn't mean to update the entire page state, just the dialog. Like, click "restart services", show the error message which one failed, and show the remaining ones which still need to be restarted. Then the user can still choose to reboot instead, or ignore the failed ones.

Reply:
So the problem here is in that case the "RestartUnit" doesn't return an error. You can see the problem when you run "systemctl status ..." saying there was a problem starting the binary, but the restart itself went fine and did not return an error.
The only times I managed to actually crash the restart was to remove the unit file itself or introduce some syntax error into it... But then tracer doesn't see it as a service anymore, but just a running process...
So generally I agree with calling tracer after a failure, but since it seems to me such a corner case I don't if it really needs a test.

Reply:
Just a quick one (and I don't see your attempt, so can't suggest anything precise): I suggest to try that locally in a VM, outside of all the test stuff. Make sure that you can modify a unit, `daemon-reload`, and `systemctl restart` then fails. Does your ExecStart= command fail properly? (/bin/false or so)?

Reply:
Oh, you are talking about the `RestartUnit` D-Bus API, sorry. That is indeed asynchronous -- you need to follow the returned `Job` objects to get their state. If that is too complicated, feel free to just use cockpit.spawn() with `systemctl restart`, that is synchronous.